src/transformers/models/t5/modeling_t5.py

```diff
@@ -773,41 +784,34 @@ def __init__(self, config, embed_tokens=None):
         self.dropout = nn.Dropout(config.dropout_rate)
 
         self.init_weights()
-        # Model parallel
+
         self.model_parallel = False
-        self.device_map = None
+        self.first_device = "cpu"
+        self.last_device = "cpu"
 
     @add_start_docstrings(PARALLELIZE_DOCSTRING)
     def parallelize(self, device_map=None):
-        # Check validity of device_map
-        self.device_map = (
-            get_device_map(len(self.block), range(torch.cuda.device_count())) if device_map is None else device_map
-        )
-        assert_device_map(self.device_map, len(self.block))
-        self.model_parallel = True
-        self.first_device = "cpu" if "cpu" in self.device_map.keys() else "cuda:" + str(min(self.device_map.keys()))
-        self.last_device = "cuda:" + str(max(self.device_map.keys()))
+        device_map = init_device_map(len(self.block), device_map)
+        self.first_device = f"cuda:{ list(device_map.keys())[0] }"
+        self.last_device = f"cuda:{ list(device_map.keys())[-1] }"
```

Comment:
Maybe we should instead put some additional work into get_device_map so it just infers based `type(model)`:
```
self.device_map = get_device_map(model) if device_map is None else device_map
validate_device_map(self.device_map)
```

And keep the old language on `first_device` and `last_device` so we can support putting part of the model on "cpu".

```
self.first_device = "cpu" if "cpu" in self.device_map.keys() else "cuda:" + str(min(self.device_map.keys()))
self.last_device = "cuda:" + str(max(self.device_map.keys())) 
```

Reply:
I'm not even sure we even need first/last devices.

But yes, I think eventually each model will have to implement its own default device map. see the work I started in https://github.com/huggingface/transformers/pull/9384 `make_default_device_map` in `model_parallel_utils.py`

So there will be no need to infer, since the model will just have a method that will do that.

Reply:
No strong opposition to update from `get_device_map` to `init_device_map`, but it would be nice to keep CPU support here if possible.

Reply:
it's not a rename, it's a change in functionality - where the default and check are merged into one function to avoid pointless code repetition - doesn't contribute to easier understanding.

and yes, point taken on supporting cpu.

I propose the first stage is to polish the API for gpus-only and then definitely include cpu. as it'd be helpful for developing/debugging MP with just one gpu.

I will start making todo items.